### ISSUE TITLE
Cypress integration test using docker-compose passes locally but fails in GitLab CI with docker:dind

### ISSUE THREAD

**Issue by briancaffey** [cypress-io/cypress#4867]
### Current behavior:

A job in my pipeline uses docker-compose with Cypress to run integration tests. The `docker-compose.ci.yml` file contains the following services:

- `postgres`
- `backend` # django webserver
- `asgiserver` # django channels daphne server
- `nginx` # serves Vue application that makes websocket requests
- `redis`

These containers are built and run as described in the `integration-tests.sh` (see below) and then start the `cypress` service with docker-compose which runs integration tests.

Out of 6 Cypress integration test, 4 passes locally. One test seems to fail sometimes and pass with the same code on retries. [Here](https://gitlab.com/verbose-equals-true/django-postgres-vue-gitlab-ecs/pipelines/73538511/builds) is an example of this behavior. The pipeline is passing, but this is only so I can capture the video output, otherwise GitLab CI will not capture the test recordings as job artifacts. The last test that tests websocket behavior fails consistently (but passes when running in docker-compose locally). 

Here is a description of my cypress tests:

* `test_api.js`: tests `cy.request` can reach an API endpoint that returns a static response
* `test_homepage.js`: tests that cypress can access the Vue application
* `test_login.js`: tests login functionality (tests the connection to the `postgres` database container) (with the same code, this test fails sometimes and passes at other times).
* `test_logout.js`: tests a user can logout; uses custom `login` Command instead of login manually as in `test_login.js` (always passes)
* `test_ws_connection.js`: presses a button that sends a websocket message and tests that the websocket response from the server triggers the creation of a notification element in the DOM. (this test fails in GitLab CI, but passes locally when running `integration-tests.sh`)
* `test_redis_connection`: sets a value to the redis cache and reads the value (this test passes)

Here is the repo: https://gitlab.com/verbose-equals-true/django-postgres-vue-gitlab-ecs

### Desired behavior:

All Cypress tests running in docker-compose should pass in GitLab CI as they pass when running locally on docker-compose. When I run docker-compose locally, I see the following logs:

```
asgiserver    | 192.168.160.6:42158 - - [29/Jul/2019:04:10:29] "WSCONNECTING /ws/ping-pong/" - -
asgiserver    | specific.cHSRkyiZ!lrvzSAdPrLsn
asgiserver    | 192.168.160.6:42158 - - [29/Jul/2019:04:10:29] "WSCONNECT /ws/ping-pong/" - -
asgiserver    | received message
asgiserver    | sending ws response
    ✓ User sees PONG message from websocket PING message (721ms)
cypress       | 
```

### Steps to reproduce: (app code and test code)

To run the tests locally with docker-compose, you can run `./integration-tests.sh` that can be found in the root of the project repository. This requires that you have docker-compose installed locally. To see the behavior in GitLab CI, you can clone the project and run the pipeline on a public shared runner. 

`integration-tests.sh` contains docker-compose commands:

```
#!/bin/bash

# set -e

echo "Starting services"
docker-compose -f docker-compose.ci.yml up -d --build

sleep 20
echo "Services are up and ready"

echo "Seeding database with user"
# docker-compose -f docker-compose.ci.yml exec backend python manage.py create_default_user

docker-compose -f docker-compose.ci.yml -f cypress.yml up --exit-code-from cypress

echo "Cypress tests passed successfully."

echo "Stopping docker compose..."
docker-compose -f docker-compose.ci.yml -f cypress.yml down
```

### Versions

Versions are documented in `docker-compose.ci.yml` and `cypress.yml`. Running the tests locally, I am on Docker version 18.09.7 with Ubuntu 16.04.

**Comment by jennifer-shehane** [comment]
I see that you are using version 3.2.0 of Cypress. Before I dive in to completely recreate the issue, could you update to the [current version of Cypress](https://on.cypress.io/changelog) and let me know if this is still happening for you? Your issue may have already been fixed - we had several performance improvements since then. Thanks!

**Comment by briancaffey** [comment]
@jennifer-shehane OK, I updated to Cypress version 3.4.0 by changing to `cypress/included:3.4.0`. This change alone caused the websocket test to fail (it passes with the 3.2.0 image). I added `cy.wait(500);` to my test script and it passed locally, but the test is still failing in GitLab CI. Here's the failing test, maybe there is something wrong with how I'm writing it:

```js
describe('Test websockets', function() {
  it("User sees PONG message from websocket PING message", function() {
    cy.login(); // get JWT
    cy.visit("/examples/websockets") // navigate to a page that opens a websocket connection
    cy.wait(500);
    cy.get("#ping").click(); // click a button that fires a message over websockets ("PING")
    cy.wait(500);
    cy.get('.pong').should('have.length', 1); // expect to see a div that is created in response to the above websocket event ("PONG")
  });
});
```

Here are the relevant GitLab CI logs from the failing test (from docker-compose with `docker:dind`):

```
cypress       |   Running: test_ws_connection.js...                                                        (6 of 6) 
cypress       | 
cypress       | 
cypress       |   Test websockets
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:48 +0000] "POST /api/auth/obtain_token/ HTTP/1.1" 200 438 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:48 +0000] "GET /examples/websockets HTTP/1.1" 200 2454 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:48 +0000] "GET /css/app.e117512b.css HTTP/1.1" 200 268424 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:48 +0000] "GET /js/app.4ffba326.js HTTP/1.1" 200 19615 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:48 +0000] "GET /js/vendor.b46ce879.js HTTP/1.1" 200 629431 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /css/701a639c.1c285fc2.css HTTP/1.1" 200 233 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d0a34f9.e3900871.js HTTP/1.1" 200 1472 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d0abad9.42cf17ba.js HTTP/1.1" 200 304 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d0b66f7.b76d307a.js HTTP/1.1" 200 416 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d0da96f.c6047424.js HTTP/1.1" 200 1048 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d0dd014.daa8bd7a.js HTTP/1.1" 200 2228 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d0e8be2.feb6f7cc.js HTTP/1.1" 200 356 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d207d33.f6123fc4.js HTTP/1.1" 200 500 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/2d22d072.b8000742.js HTTP/1.1" 200 801 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/4b47640d.83b94a0a.js HTTP/1.1" 200 12364 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:49 +0000] "GET /js/701a639c.77289ef5.js HTTP/1.1" 200 639 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:50 +0000] "GET /api/users/profile/ HTTP/1.1" 200 72 "http://nginx/examples/websockets" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:50 +0000] "GET /fonts/flUhRq6tzZclQEJ-Vdg-IuiaDsNcIhQ8tQ.0509ab09.woff2 HTTP/1.1" 200 60840 "http://nginx/css/app.e117512b.css" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:50 +0000] "GET /fonts/KFOmCnqEu92Fr1Mu4mxM.60fa3c06.woff HTTP/1.1" 200 20268 "http://nginx/css/app.e117512b.css" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:50 +0000] "GET /fonts/KFOlCnqEu92Fr1MmEU9fBBc-.87284894.woff HTTP/1.1" 200 20464 "http://nginx/css/app.e117512b.css" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
asgiserver    | 172.19.0.6:44884 - - [30/Jul/2019:18:12:52] "WSCONNECTING /ws/ping-pong/" - -
asgiserver    | specific.MhARSrCQ!CaVePvbSUEYD
asgiserver    | 172.19.0.6:44884 - - [30/Jul/2019:18:12:53] "WSCONNECT /ws/ping-pong/" - -
cypress       | 
    1) User sees PONG message from websocket PING message
cypress       | 
cypress       | 
cypress       |   0 passing (9s)
cypress       |   1 failing
cypress       | 
cypress       |   1) Test websockets User sees PONG message from websocket PING message:
cypress       |      CypressError: Timed out retrying: Expected to find element: '.pong', but never found it.
cypress       |       at Object.cypressErr (http://nginx/__cypress/runner/cypress_runner.js:84963:11)
cypress       |       at Object.throwErr (http://nginx/__cypress/runner/cypress_runner.js:84916:18)
cypress       |       at Object.throwErrByPath (http://nginx/__cypress/runner/cypress_runner.js:84947:17)
cypress       |       at retry (http://nginx/__cypress/runner/cypress_runner.js:77960:16)
cypress       |       at http://nginx/__cypress/runner/cypress_runner.js:69765:18
cypress       |       at tryCatcher (http://nginx/__cypress/runner/cypress_runner.js:134216:23)
cypress       |       at Promise._settlePromiseFromHandler (http://nginx/__cypress/runner/cypress_runner.js:132234:31)
cypress       |       at Promise._settlePromise (http://nginx/__cypress/runner/cypress_runner.js:132291:18)
cypress       |       at Promise._settlePromise0 (http://nginx/__cypress/runner/cypress_runner.js:132336:10)
cypress       |       at Promise._settlePromises (http://nginx/__cypress/runner/cypress_runner.js:132411:18)
cypress       |       at Async._drainQueue (http://nginx/__cypress/runner/cypress_runner.js:129140:16)
cypress       |       at Async._drainQueues (http://nginx/__cypress/runner/cypress_runner.js:129150:10)
cypress       |       at Async.drainQueues (http://nginx/__cypress/runner/cypress_runner.js:129024:14)
cypress       |       at <anonymous>
cypress       | 
cypress       | 
cypress       | 
cypress       | 
cypress       |   (Results)
cypress       | 
cypress       |   ┌─────────────────────────────────────┐
cypress       |   │ Tests:        1                     │
cypress       |   │ Passing:      0                     │
cypress       |   │ Failing:      1                     │
cypress       |   │ Pending:      0                     │
cypress       |   │ Skipped:      0                     │
cypress       |   │ Screenshots:  1                     │
cypress       |   │ Video:        true                  │
cypress       |   │ Duration:     9 seconds             │
cypress       |   │ Spec Ran:     test_ws_connection.js │
cypress       |   └─────────────────────────────────────┘
cypress       | 
cypress       | 
cypress       |   (Screenshots)
cypress       | 
cypress       |   - /e2e/cypress/screenshots/test_ws_connection.js/Test websockets -- User sees PONG message from websocket PING message (failed).png (1280x720)
cypress       | 
asgiserver    | 172.19.0.6:44884 - - [30/Jul/2019:18:12:56] "WSDISCONNECT /ws/ping-pong/" - -
nginx         | 172.19.0.7 - - [30/Jul/2019:18:12:56 +0000] "GET /ws/ping-pong/ HTTP/1.1" 101 4 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.4.0 Chrome/61.0.3163.100 Electron/2.0.18 Safari/537.36"
cypress       | 
cypress       |   (Video)
cypress       | 
cypress       |   - Started processing:   Compressing to 32 CRF
cypress       |   - Finished processing:  /e2e/cypress/videos/test_ws_connection.js.mp4 (5 seconds)
cypress       | 
cypress       | 
cypress       | ====================================================================================================
cypress       | 
cypress       |   (Run Finished)
cypress       | 
cypress       | 
cypress       |       Spec                                                Tests  Passing  Failing  Pending  Skipped 
cypress       |   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
cypress       |   │ ✔ test_api.js                               00:03        1        1        -        -        - │
cypress       |   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
cypress       |   │ ✔ test_homepage.js                          00:02        1        1        -        -        - │
cypress       |   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
cypress       |   │ ✔ test_login.js                             00:05        1        1        -        -        - │
cypress       |   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
cypress       |   │ ✔ test_logout.js                            00:06        1        1        -        -        - │
cypress       |   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
cypress       |   │ ✔ test_redis_connection.js                  00:04        1        1        -        -        - │
cypress       |   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
cypress       |   │ ✖ test_ws_connection.js                     00:09        1        -        1        -        - │
cypress       |   └────────────────────────────────────────────────────────────────────────────────────────────────┘
cypress       |     1 of 6 failed (17%)                         00:31        6        5        1        -        -  
cypress       | 
cypress exited with code 1
```

**Comment by jennifer-shehane** [comment]
@briancaffey Is this still an issue for your project?

**Comment by jennifer-shehane** [comment]
Unfortunately we have to close this issue due to inactivity. Please comment if there is new information to provide concerning the original issue and we can reopen.

**Comment by joelvandiver** [comment]
For what it's worth to whoever may land here in the future, I ran into a similar problem.  

I found out that the handling of volumes in `docker-dind` didn't work the same as when I ran `docker-compose` locally.  GitLab mounts the code of job onto a volume.  I then was trying to mount into the cypress image through docker compose.

I was able to solve this (after some help from a friend) by finding the original volume mount in the GitLab CI:

> `.gitlab-ci.yml` 

```yml
  script:
    # Mount the volume through dind
    - export CONTAINER_ID=$(docker ps -q -f "label=com.gitlab.gitlab-runner.job.id=$CI_JOB_ID" -f "label=com.gitlab.gitlab-runner.type=build")
    - export MOUNT_DIR=$(docker inspect $CONTAINER_ID -f "{{ range .Mounts }}{{ if eq .Destination \"/builds\" }}{{ .Source }}{{end}}{{end}}")
```

I was then able to use the `$MOUNT_DIR` var in my docker compose for the cypress image when mounting the volume:

> `docker-compose.yml`

```yml
...
  # Cypress container
  cypress:
    # https://github.com/cypress-io/cypress-docker-images
    image: "cypress/included:9.1.0"
    depends_on:
      - api
    environment:
      - CYPRESS_baseUrl=http://api:8000
    working_dir: /e2e
    volumes:
      - ${MOUNT_DIR}/:/e2e
    networks:
      api-int:
...
```

I hope this helps!
